packages/core/src/compiler/Substance.ts

```diff
@@ -326,37 +396,609 @@ const checkStmt = (stmt: SubStmt<A>, env: Env): CheckerResult<SubStmt<A>> => {
+    else if (operator === ">=")
+      return ok(closeEqual(lVal, rVal) ? true : lVal >= rVal);
+    else if (operator === "==") return ok(closeEqual(lVal, rVal));
+    else return ok(!closeEqual(lVal, rVal));
+  }
+};
+
+const closeEqual = (x: number, y: number): boolean => {
+  const EPS = 0.00001;
+  return Math.abs(x - y) < EPS;
+};
+
+const evalNum = (
+  n: NumExpr<A>,
+  subst: ISetSubst,
+): Result<number, SubstanceError> => {
+  const result = evalNumHelper(n, subst);
+  if (result.isErr()) return err(result.error);
+
+  const value = result.value;
+
+  // This is the case of NaN
+  if (value !== value) {
+    return err({
+      tag: "InvalidArithmeticValueError",
+      location: n,
+      value,
+    });
+  }
+  return ok(value);
+};
+
+const evalNumHelper = (
+  n: NumExpr<A>,
+  subst: ISetSubst,
+): Result<number, SubstanceError> => {
+  if (n.tag === "NumberConstant") {
+    return ok(n.value);
+  } else if (n.tag === "Identifier") {
+    return substISetVarNumber(n.value, n, subst);
+  } else if (n.tag === "UnaryExpr") {
+    const { arg } = n;
+    const argValRes = evalNum(arg, subst);
+    return argValRes.andThen((n) => ok(-n));
+  } else {
+    const { operator, left, right } = n;
+    const lValRes = evalNum(left, subst);
+    if (lValRes.isErr()) return err(lValRes.error);
+    const rValRes = evalNum(right, subst);
+    if (rValRes.isErr()) return err(rValRes.error);
+
+    const lVal = lValRes.value,
+      rVal = rValRes.value;
+
+    if (operator === "+") return ok(lVal + rVal);
+    else if (operator === "-") return ok(lVal - rVal);
+    else if (operator === "*") return ok(lVal * rVal);
+    else if (operator === "^") return ok(lVal ** rVal);
+    else {
+      // div or mod
+      if (rVal === 0) {
+        return err({
+          tag: "DivideByZeroError",
+          location: n,
+        });
+      }
+      if (operator === "/") return ok(lVal / rVal);
+      else return ok(lVal % rVal);
+    }
+  }
+};
+
+const substISetVarNumber = (
+  v: string,
+  location: AbstractNode,
+  subst: ISetSubst,
+): Result<number, SubstanceError> => {
+  // If already a number, use that number.
+  if (!isNaN(Number(v))) {
+    return ok(Number(v));
+  }
+
+  const isetVarValue = subst.get(v);
+  if (isetVarValue === undefined) {
+    return err({
+      tag: "InvalidSetIndexingError",
+      index: v,
+      location,
+      suggestions: [...subst.keys()],
+    });
+  }
+  return ok(isetVarValue);
+};
+
+const substISetVarStr = (
+  v: string,
+  location: AbstractNode,
+  subst: ISetSubst,
+): Result<string, SubstanceError> => {
+  const underscorePos = v.lastIndexOf("_");
+  if (underscorePos === -1) {
+    return ok(v);
+  }
+
+  const prefix = v.slice(0, underscorePos);
+  const isetVarName = v.slice(underscorePos + 1);
+
+  const isetVarValue = substISetVarNumber(isetVarName, location, subst);
+  if (isetVarValue.isErr()) return err(isetVarValue.error);
+  return ok(`${prefix}_${isetVarValue.value}`);
```

Comment:
```suggestion
  return andThen((idx) => ok(`${prefix}_${idx}`), isetVarValue);
```

Reply:
What's the difference between `andThen(..., isetVarValue)` and `isetVarValue.andThen(...)`? I find the latter to be more intuitive.

Reply:
never knew that existed! if that works i agree it’s more intuitive for sure

Reply:
yep it works